joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/ConcurrentTaskExecutionUtil.scala

```diff
@@ -0,0 +1,54 @@
+package io.joern.x2cpg.utils
+
+import io.shiftleft.utils.ExecutionContextProvider
+
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.Try
+
+/** Following ConcurrentWriterCpgPass, this creates a generic re-usable utility for bootstrapping basic non-CPG pass
+  * concurrency.
+  */
+object ConcurrentTaskExecutionUtil {
+
+  private val MAX_QUEUE_SIZE_DEFAULT = 2 + 4 * Runtime.getRuntime.availableProcessors()
+
+  /** Uses the parallel queue strategy from [[io.shiftleft.passes.ConcurrentWriterCpgPass]] to offer a generic means of
+    * executing an iterator of tasks that share an output type in parallel.
+    *
+    * @param tasks
+    *   the tasks to parallelize.
+    * @param maxQueueSize
+    *   the max number of tasks to queue for parallel execution.
+    * @tparam V
+    *   the output type of each task.
+    * @return
+    *   an array of the executed tasks as either a success or failure.
+    * @see
+    *   [[io.shiftleft.passes.ConcurrentWriterCpgPass]]
+    */
+  def runInParallel[V](tasks: Iterator[() => V], maxQueueSize: Int = MAX_QUEUE_SIZE_DEFAULT)(implicit
+    ec: ExecutionContext = ExecutionContextProvider.getExecutionContext
+  ): List[Try[V]] = {
+    val completionQueue = mutable.ArrayDeque.empty[Future[V]]
+    val results         = mutable.ArrayBuffer.empty[Try[V]]
+
+    var done = false
+    while (!done) {
+      if (completionQueue.size < maxQueueSize && tasks.hasNext) {
+        val nextTask = tasks.next()
+        completionQueue.append(Future.apply(nextTask()))
+      } else if (completionQueue.nonEmpty) {
+        val future = completionQueue.removeHead()
+        val res    = Try(Await.result(future, Duration.Inf))
+        results.append(res)
+      } else {
+        done = true
+      }
+    }
+
+    results.toList
```

Comment:
```suggestion
    val futures = tasks.map(task => Future(task())).toSeq
    // starts them all ^

    futures.map { future =>
      Try(Await.result(future, Duration.Inf))
    }
```

Reply:
doesn't this do the same with less noise?

Reply:
The point of the queue limits how many tasks are loaded into memory - which is also helped along with the use of the iterator parameter. Not sure, on the other hand, if it makes that much of an impact using this pattern outside of CPG writing to disk.

Reply:
I can offer this third way of doing things from one of my joern scripts:

```scala

def runInParallel[V](it: Iterator[() => V]): mutable.Buffer[V] = {
    StreamSupport
      .stream(Spliterators.spliteratorUnknownSize(it.asJava, Spliterator.NONNULL), /* parallel */ true)
      .map(_.apply())
      .toList
      .asScala
  }
```

I haven't benchmarked it though - I just assumed the Java streams are optimized well enough (including lazy reading from the stream/iterator)

Reply:
I guess both @maltek and my suggestions ignore the fact that @DavidBakerEffendi wants to run `x` tasks in parallel - our suggestions both start them all eagerly. 

So I guess it's best to just ignore us. 
Final thought: doesn't ExecutorService with a fixed threadpool do that as well?
But again, feel free to ignore, we didn't mean to hold you up, it was just a thought.

Reply:
Got back from vacation today, I'll see if I can use @maltek's solution can be specified with a fixed `x` size as it looks much cleaner and Java's way of doing things are likely more battle-tested.

Yeah, my idea was that I don't want to pollute the memory too much for tasks which may produce unpredictable memory demands. Let me see if I can produce a cleaner variant that addresses this, and maybe reflect the intention of the method in the name.

Reply:
Seems like a [ThreadPoolExecutor](https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/ThreadPoolExecutor.html) is what I want, should limit the number of running tasks.

Reply:
yes that's what I had in mind re ExecutorService

you get bonus points if you wrap it in a scala.util.Using - that'll ensure that the executorservice is always properly closed, similar to java's try-with-resource